File: Entregable 2/cuatro.rb
if ARGV.count < 5
  puts "Missing arguments"
  puts "\t * N cantidad de elementos del vector"
  puts "\t * MAX_NUM numero maximo de cada elemento del vector"
  puts "\t * X elemento a buscar ocurrencias"
  puts "\t * T cantidad de threads"
  puts "\t * I imprime?"
  return
end

N = ARGV[0].to_i
MAX_NUM = ARGV[1].to_i
X = ARGV[2].to_i
T = ARGV[3].to_i
imprimir = ARGV[4].to_i

puts "\t * #{N} cantidad de elementos del vector"
puts "\t * #{MAX_NUM} numero maximo de cada elemento del vector"
puts "\t * #{X} elemento a buscar ocurrencias"
puts "\t * #{T} cantidad de threads"
puts "\t * #{imprimir} imprime?"

if N < T
  puts "La cantidad de elementos no puede ser menor a la cantidad de threads"
  return
end

@ocurrencias = 0
@ocurrencias_vector = []
@vector = []

(1..N).each do
  @vector.push(rand(MAX_NUM))
end

puts "#{@vector.map{|v| v}}" if imprimir

def contar_ocurrencias(index)
  last_position = index*(N/T) + ((N/T)-1)
  first_position = index*(N/T)
  for i in first_position..last_position
    if (@vector[i] == X)
      @ocurrencias_vector[index] = @ocurrencias_vector[index] + 1
    end
  end

  if index == T-1
    @ocurrencias_vector.each do |conteo_parcial|
      @ocurrencias = @ocurrencias + conteo_parcial
    end
  end
end

threads = []
START = Time.now
(0..T-1).each do |thread_number|
  @ocurrencias_vector[thread_number] = 0
  threads << Thread.new { contar_ocurrencias(thread_number) }
end

threads.each do |thread|
  thread.join
end

puts "...Tiempo obtenido: #{Time.now - START}"
puts "Ocurrencias obtenidas -> #{@ocurrencias}"

threads.each do |thread|
  thread.exit
end

#TODO : Not working as expected.. debug it

File: Entregable 2/cuatro.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <omp.h>

double dwalltime();
void imprimeVector(uint64_t *V, uint64_t size);

int main(int argc,char*argv[]){
  //Controla los argumentos al programa
  if (argc < 5){
    printf("\n Falta un argumento : \n");
    printf("\t * N cantidad de elementos del vector\n");
    printf("\t * MAX_NUM numero maximo de cada elemento del vector\n");
    printf("\t * X elemento a buscar ocurrencias \n");
    printf("\t * T cantidad de threads \n");
    printf("\t * I imprime?\n");
    return 1;
  }

  uint64_t ocurrencias = 0;
  int I;
  double timetick;
  uint64_t N, M, X, T, i;
  uint64_t *vector;

  N = atoi(argv[1]);
  M = atoi(argv[2]);
  X = atoi(argv[3]);
  T = atoi(argv[4]);
  I = atoi(argv[5]);
  omp_set_num_threads(T);

  if (N < T) {
    printf("Cantidad de elementos no puede ser menor a la cantidad de threads");
    return 1;
  }

  // Genero el vector de N elementos con numero aleatorios
  srand(time(NULL));
  vector = (uint64_t*)malloc(sizeof(uint64_t)*N);
  for (i=0; i<N; i++) {
    vector[i] = (rand() % M) + 1;
  }

  printf("\t * %llu : cantidad de elementos del vector\n", N);
  printf("\t * %llu : numero maximo de cada elemento del vector\n", M);
  printf("\t * %llu : elemento a buscar ocurrencias \n", X);
  printf("\t * %llu : cantidad de threads \n", T);
  if ( I == 1 )
    imprimeVector(vector, N);

  timetick = dwalltime();
  #pragma omp parallel for reduction(+ : ocurrencias)
    for (i=0; i<N; i++) {
      if (vector[i] == X)
        ocurrencias++;
    }

  printf("Tiempo en segundos %f \n", dwalltime() - timetick);
  printf("\n\t Se encontraron %llu ocurrencias\n", ocurrencias);

  return 0;
}

/*****************************************************************/

#include <sys/time.h>

double dwalltime()
{
  double sec;
  struct timeval tv;

  gettimeofday(&tv,NULL);
  sec = tv.tv_sec + tv.tv_usec/1000000.0;
  return sec;
}

void imprimeVector(uint64_t *V, uint64_t size){
  printf("Imprimiendo matriz...\n");
  uint64_t i;
  printf("[");
  for (i=0; i<size; i++) {
    if (i==size-1)
      printf("%llu]\n", V[i]);
    else
      printf("%llu, ", V[i]);
  }
}
